### PR TITLE
bal 2254 rule engine v2

### PR DESCRIPTION
**Pull request by liorzam**
image

## Summary by CodeRabbit

- **New Features**
	- Introduced a Rule Engine module to the application, enhancing rule validation and error handling capabilities.
	- Added custom error handling for rule-based operations.
	- Implemented various comparison operators for comprehensive rule evaluations.
	- Provided type definitions and schemas for structured rule processing.

- **Tests**
	- Included unit tests to ensure the accuracy and reliability of rule validations.

### CODE REVIEW

File: services/workflows-service/src/rule-engine/core/operators/constants.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import { OPERATION } from './enums';
+
+export const operationHelpers = {
+  [OPERATION.EQUALS]: EQUALS,
+  [OPERATION.BETWEEN]: BETWEEN,
+  [OPERATION.GT]: GT,
+  [OPERATION.GTE]: GTE,
+  [OPERATION.LT]: LT,
+  [OPERATION.LTE]: LTE,
+  [OPERATION.LAST_YEAR]: LAST_YEAR,
```
Comment (Blokh) [inline services/workflows-service/src/rule-engine/core/operators/constants.ts:12]: Why is last year a a part of helper operations?

File: services/workflows-service/src/rule-engine/core/operators/enums.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+export const OPERATION = {
+  EQUALS: 'EQUALS',
+  BETWEEN: 'BETWEEN',
+  GT: 'GT',
+  LT: 'LT',
+  GTE: 'GTE',
+  LTE: 'LTE',
+  LAST_YEAR: 'LAST_YEAR',
```
Comment (Blokh) [inline services/workflows-service/src/rule-engine/core/operators/enums.ts:8]: this should not be here

File: services/workflows-service/src/rule-engine/core/operators/helpers.ts [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+  eval: ConditionFn<primitive> = (dataValue: primitive, conditionValue: primitive): boolean => {
+    return (
+      this.equals.execute(dataValue, conditionValue) ||
+      this.lessThan.execute(dataValue, conditionValue)
+    );
+  };
+}
+
+class Between extends BaseOperator<BetweenParams> {
```
Comment (Blokh) [inline services/workflows-service/src/rule-engine/core/operators/helpers.ts:142]: between should be by 2 operators, before x and after y.

File: services/workflows-service/src/rule-engine/core/operators/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    super({
+      operator: 'BETWEEN',
+      conditionValueSchema: BetweenSchema,
+      dataValueSchema: PrimitiveSchema,
+    });
+    this.gte = new GreaterThanOrEqual();
+    this.lte = new LessThanOrEqual();
+  }
+
+  eval: ConditionFn<BetweenParams> = (
```
Comment (Blokh) [inline services/workflows-service/src/rule-engine/core/operators/helpers.ts:156]: eval is the functionality of one of the conversions of the values.
evaluate is a rule valuation function

File: services/workflows-service/src/rule-engine/core/rule-engine.unit.test.ts [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+      ],
+    };
+
+    const result = runRuleSet(ruleSetExample, mockData);
+    expect(result).toBeDefined();
+    expect(result).toHaveLength(1);
+    expect(result[0]).toMatchObject({
+      error: expect.any(MissingKeyError),
+      message: 'Rule is missing the key field',
+      passed: false,
```
Comment (liorzam) [inline services/workflows-service/src/rule-engine/core/rule-engine.unit.test.ts:186]: remove

**Pull request merged**